### DISCUSSION TITLE
Auto reconnect subscription after CLOSED connection

### DISCUSSION THREAD

**Discussion by farzd** [supabase/supabase#27513] [Questions]
I have the following reconnect strategy for my React Native iOS app. Testing in production 
```
"expo": "^50.0.17",
"@supabase/supabase-js": "^2.42.5"
```
When I put the app into background - after 3 seconds the subscription disconnects ```CHANNEL_ERROR``` - this is expected behaviour. When the app is brought to the foreground, the subscriptions automatically attempts to reconnect ```SUBSCRIBED``` - this works great as well.
There are some instances where subscription fails to reconnect and finally sends the status ```CLOSED``` and finally ```TIMED_OUT```

I then attempt a reconnect with the code below however this appears to be stuck in a loop, my logs show that the status fluctuates between ```SUBSCRIBED``` and ```CLOSED```. When i force a manual disconnect using ``` supabase.removeChannel(XXX)``` 
my reconnect code works but when the connection fails for other reasons, the reconnect get stuck in a strange loop.

```
 const subscriptionRef = useRef(null)
  const reconnectAttemptsRef = useRef(0)

  useEffect(() => {
    if (user_id) {
      subscribeToChannel()
    }

    return () => {
      if (subscriptionRef.current) {
        supabase.removeChannel(subscriptionRef.current)
      }
    }
  }, [user_id])

  const subscribeToChannel = () => {
    subscriptionRef.current = supabase
      .channel(`XXX:${user_id}`)
      .on(
        'postgres_changes',
        {
          event: '*',
          schema: 'public',
          table: 'table',
          filter: `user_id=eq.${user_id}`,
        },
        (payload) => {
          // Reset reconnect attempts on successful message
          reconnectAttemptsRef.current = 0;
        }
      )
      .subscribe((status) =>  {
        console.log({ status });

        if (status === "SUBSCRIBED") {
          console.log("Successfully subscribed");
          reconnectAttemptsRef.current = 0;

        } else if (status === "CLOSED") {
          console.log("Subscription closed");
          reconnectWithBackoff();
        }
      })
  }

  const reconnectWithBackoff = () => {
    if (
      reconnectAttemptsRef.current < 5
    ) {
      const delay = 3000 * Math.pow(2, reconnectAttemptsRef.current)
      console.log(`Reconnecting in ${delay}ms...`)

      setTimeout(() => {
        console.log(`Reconnect attempt ${reconnectAttemptsRef.current + 1}`)
        subscribeToChannel()
        reconnectAttemptsRef.current += 1
      }, delay)
    } else {
      console.log('Max reconnection attempts reached. Please try again later.')
    }
  }

```

Only other time i can recreate this is with fast refresh enabled on the iOS simulator

**Comment by farzd** [comment]
![IMG_1697](https://github.com/supabase/supabase/assets/1423413/147c381e-2499-4d87-bc56-628a43b15331)
logs printed to screen from production app on iOS native device

**Reply by GaryAustin1** [reply]
Here are two studies I did on realtime and reconnecting.  They show process that worked for me at the time.  Light testing on mobile browsers but not a native app.
https://github.com/GaryAustin1/Realtime2/blob/main/src/app.js
https://github.com/orgs/supabase/discussions/5641
